[Wasm-GC] Introduce array types

takikawa opened the pull request:
#### fa3d441596d43e14b9ce4b6dff3b38e9d7b81bc3
<pre>
[Wasm-GC] Introduce array types
<a href="https://bugs.webkit.org/show_bug.cgi?id=240980">https://bugs.webkit.org/show_bug.cgi?id=240980</a>

Reviewed by Justin Michaud.

Adds support for creating type definitions for Wasm GC array types.
Also adds support for the `arrayref` type, which is the supertype of all
array references. The patch does not yet add support for operations
which create arrays (except for the null array reference) or access
them.

* JSTests/wasm/gc/arrays.js: Added.
(module):
(testArrayDeclaration):
* JSTests/wasm/gc/rec.js:
(testRecDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isArrayref):
(JSC::Wasm::isRefWithTypeIndex):
(JSC::Wasm::isSubtype):
(JSC::Wasm::isValidHeapTypeKind):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser&lt;SuccessType&gt;::parseValueType):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseStructType):
(JSC::Wasm::SectionParser::parseArrayType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeDefinition::dump const):
(JSC::Wasm::ArrayType::toString const):
(JSC::Wasm::ArrayType::dump const):
(JSC::Wasm::computeStructTypeHash):
(JSC::Wasm::computeArrayTypeHash):
(JSC::Wasm::TypeDefinition::hash const):
(JSC::Wasm::TypeDefinition::tryCreateArrayType):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
(JSC::Wasm::ArrayParameterTypes::hash):
(JSC::Wasm::ArrayParameterTypes::equal):
(JSC::Wasm::ArrayParameterTypes::translate):
(JSC::Wasm::TypeInformation::typeDefinitionForStruct):
(JSC::Wasm::TypeInformation::typeDefinitionForArray):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::FieldType::operator== const):
(JSC::Wasm::FieldType::operator!= const):
(JSC::Wasm::StructType::StructType):
(JSC::Wasm::StructType::field const):
(JSC::Wasm::StructType::getField):
(JSC::Wasm::StructType::storage):
(JSC::Wasm::StructType::storage const):
(JSC::Wasm::ArrayType::ArrayType):
(JSC::Wasm::ArrayType::elementType const):
(JSC::Wasm::ArrayType::getElementType):
(JSC::Wasm::ArrayType::storage):
(JSC::Wasm::ArrayType::storage const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedStructSize):
(JSC::Wasm::TypeDefinition::allocatedArraySize):
(JSC::Wasm::StructField::operator== const): Deleted.
(JSC::Wasm::StructField::operator!= const): Deleted.
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/254455@main">https://commits.webkit.org/254455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/01409749923521ef3fb92d860a4da1bc24f768a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98288 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154649 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32067 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27652 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81366 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92810 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94638 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25462 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75962 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25394 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68363 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/80712 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29858 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14393 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/74500 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29589 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15370 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26302 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3120 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38312 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/77361 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31718 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34465 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17136 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFormat.h:
```diff
@@ -122,6 +122,12 @@ inline bool isI31ref(Type type)
 
+inline bool isArrayref(Type type)
+{
+    ASSERT(Options::useWebAssemblyGC());
```
release assert for now?

justinmichaud replied:
Hmm, is this assert even valid? We could have typed function references but no gc, and this will be called?

takikawa replied:
Yes, I think you're right that this is incorrect. I think the assertion in the I31ref function above that is incorrect too, and causes some issues in a debug test run for the `call_ref` tests (because typed function refs are turned on, but GC is turned off).

Instead, it should probably always return `false` when GC is off. Edit: I made this change in the latest push.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -117,16 +134,31 @@ static unsigned computeStructTypeHash(size_t fieldCount, const StructField* fiel
+    unsigned accumulator = 0x7835ab;
+    accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<uint8_t>::hash(static_cast<uint8_t>(elementType.type.kind)));
+    accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<uint8_t>::hash(static_cast<uint8_t>(elementType.type.nullable)));
+    accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<TypeIndex>::hash(elementType.type.index));
```
Are type indexes unique? As in, the type index is correlated with kind and mutability? Why not just hash type index + module?

takikawa replied:
In this case no, it's not unique. In particular, the null type index is used for a lot of types. i.e., all non-reftypes like `I32`. Even if the type index is non-null, the mutability can vary and the kind can be either `Ref` or `RefNull`.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -92,58 +92,93 @@ enum Mutability : uint8_t {
     static size_t allocatedFunctionSize(Checked<FunctionArgCount> retCount, Checked<FunctionArgCount> argCount) { return sizeof(TypeDefinition) + (retCount + argCount) * sizeof(Type); }
-    static size_t allocatedStructSize(Checked<StructFieldCount> fieldCount) { return sizeof(TypeDefinition) + fieldCount * sizeof(StructField); }
+    static size_t allocatedStructSize(Checked<StructFieldCount> fieldCount) { return sizeof(TypeDefinition) + fieldCount * sizeof(FieldType); }
+    static size_t allocatedArraySize() { return sizeof(TypeDefinition) + sizeof(FieldType); }
```
This is just the length + the pointer to the backing store?

takikawa replied:
For arrays, it's the pointer to the backing store and also the FieldType. Though it's slightly larger from the variant type taking up more than the pointer.

Looking at this again, I'm wondering if it's worth having `ArrayType` use a payload pointer since it doesn't have a variable length of data to store. It possibly saves a small amount of space in the `m_typeHeader` in `TypeDefinition` because a `FieldType` is slightly bigger than the size of a `FuncSignature`, so having `ArrayType` store a `FieldType` directly would grow the `std::variant` size slightly.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -260,6 +303,39 @@ struct StructParameterTypes {
+        return true;
+    }
+
+    static void translate(TypeHash& entry, const ArrayParameterTypes& params, unsigned)
```
What does this function do?

takikawa replied:
It's used to actually construct the hash set entry when it's needed, i.e., if the hash set has determined that the entry is not already in the set. If it's already in the set, the user of the set can avoid constructing the entry.

(relevant source code comment https://searchfox.org/wubkat/source/Source/WTF/wtf/HashSet.h#105)

justinmichaud reviewed (commented):
I left a few questions

justinmichaud reviewed (approved):
r=me

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -514,6 +589,15 @@ RefPtr<TypeDefinition> TypeInformation::typeDefinitionForStruct(const Vector<Str
+    Locker locker { info.m_lock };
+
+    auto addResult = info.m_typeSet.template add<ArrayParameterTypes>(ArrayParameterTypes { elementType });
+    return addResult.iterator->key;
```
Is this iterator thread-safe? What happens if the backing store is resized?

takikawa replied:
I think this use and similar uses above should be safe, because it's protected by a `Locker` in this scope, and the iterator itself doesn't escape. The key is stored in the `TypeHash` that underlies the iterator, but its lifetime is not tied to the iterator (because it's malloced in the `tryCreateX` functions that are called from the various `translate` methods) so it's ok if it is returned here.

justinmichaud reviewed (approved):
r=me again

takikawa committed:
[Wasm-GC] Introduce array types
https://bugs.webkit.org/show_bug.cgi?id=240980

Reviewed by Justin Michaud.

Adds support for creating type definitions for Wasm GC array types.
Also adds support for the `arrayref` type, which is the supertype of all
array references. The patch does not yet add support for operations
which create arrays (except for the null array reference) or access
them.

* JSTests/wasm/gc/arrays.js: Added.
(module):
(testArrayDeclaration):
* JSTests/wasm/gc/rec.js:
(testRecDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isArrayref):
(JSC::Wasm::isRefWithTypeIndex):
(JSC::Wasm::isSubtype):
(JSC::Wasm::isValidHeapTypeKind):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser<SuccessType>::parseValueType):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseStructType):
(JSC::Wasm::SectionParser::parseArrayType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeDefinition::dump const):
(JSC::Wasm::ArrayType::toString const):
(JSC::Wasm::ArrayType::dump const):
(JSC::Wasm::computeStructTypeHash):
(JSC::Wasm::computeArrayTypeHash):
(JSC::Wasm::TypeDefinition::hash const):
(JSC::Wasm::TypeDefinition::tryCreateArrayType):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
(JSC::Wasm::ArrayParameterTypes::hash):
(JSC::Wasm::ArrayParameterTypes::equal):
(JSC::Wasm::ArrayParameterTypes::translate):
(JSC::Wasm::TypeInformation::typeDefinitionForStruct):
(JSC::Wasm::TypeInformation::typeDefinitionForArray):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::FieldType::operator== const):
(JSC::Wasm::FieldType::operator!= const):
(JSC::Wasm::StructType::StructType):
(JSC::Wasm::StructType::field const):
(JSC::Wasm::StructType::getField):
(JSC::Wasm::StructType::storage):
(JSC::Wasm::StructType::storage const):
(JSC::Wasm::ArrayType::ArrayType):
(JSC::Wasm::ArrayType::elementType const):
(JSC::Wasm::ArrayType::getElementType):
(JSC::Wasm::ArrayType::storage):
(JSC::Wasm::ArrayType::storage const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedStructSize):
(JSC::Wasm::TypeDefinition::allocatedArraySize):
(JSC::Wasm::StructField::operator== const): Deleted.
(JSC::Wasm::StructField::operator!= const): Deleted.
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/254455@main
